File: rotom_dex/domain/conditions.py
"""Typed, fail-closed condition vocabulary with three-valued evaluation.

Conditions describe evolution requirements and acquisition prerequisites. They
are validated structurally on write and on read. `unknown` leaves carry a reason
and never evaluate to true; absent context means unknown, not false.
"""

from __future__ import annotations

# op -> ("none" | "identifier" | "int" | "choice", allowed choices)
_LEAVES: dict[str, tuple[str, tuple[str, ...]]] = {
    "always": ("none", ()),
    "trade": ("none", ()),
    "overworld_rain": ("none", ()),
    "device_upside_down": ("none", ()),
    "level_at_least": ("int", ()),
    "happiness_at_least": ("int", ()),
    "beauty_at_least": ("int", ()),
    "affection_at_least": ("int", ()),
    "has_pokemon": ("identifier", ()),
    "has_item": ("identifier", ()),
    "use_item": ("identifier", ()),
    "held_item": ("identifier", ()),
    "knows_move": ("identifier", ()),
    "knows_move_type": ("identifier", ()),
    "party_has_pokemon": ("identifier", ()),
    "party_has_type": ("identifier", ()),
    "trade_for_pokemon": ("identifier", ()),
    "at_location": ("identifier", ()),
    "in_region": ("identifier", ()),
    "milestone": ("identifier", ()),
    "encounter_condition": ("identifier", ()),
    "encounter_pokemon": ("identifier", ()),
    "time_of_day": ("identifier", ()),
    "gender": ("choice", ("male", "female")),
    "stat_relation": ("choice", ("attack>defense", "attack<defense", "attack=defense")),
}

# Context keys used by numeric leaves.
_NUMERIC_CONTEXT = {
    "level_at_least": "level",
    "happiness_at_least": "happiness",
    "beauty_at_least": "beauty",
    "affection_at_least": "affection",
}
_INT_RANGES = {
    "level_at_least": (1, 100),
    "happiness_at_least": (1, 255),
    "beauty_at_least": (1, 255),
    "affection_at_least": (1, 255),
}

SUPPORTED_OPS = frozenset({"and", "or", "unknown", *_LEAVES})


def validate_condition(value: object) -> None:
    if not isinstance(value, dict) or "op" not in value:
        raise ValueError("Condition must be an object with op")
    op = value["op"]
    if op in {"and", "or"}:
        args = value.get("args")
        if set(value) != {"op", "args"} or not isinstance(args, list) or not args:
            raise ValueError("and/or require nonempty args")
        for arg in args:
            validate_condition(arg)
        return
    if op == "unknown":
        reason = value.get("reason")
        if set(value) != {"op", "reason"} or not isinstance(reason, str) or not reason:
            raise ValueError("Unknown conditions require a reason")
        return
    if op not in _LEAVES:
        raise ValueError(f"Unsupported condition: {op}")
    kind, choices = _LEAVES[op]
    if kind == "none":
        if set(value) != {"op"}:
            raise ValueError(f"{op} takes no arguments")
        return
    if set(value) != {"op", "value"}:
        raise ValueError(f"{op} requires exactly one value")
    arg = value["value"]
    if kind == "int":
        low, high = _INT_RANGES[op]
        if type(arg) is not int or not low <= arg <= high:
            raise ValueError(f"Invalid {op} value: {arg!r}")
    elif kind == "identifier":
        if not isinstance(arg, str) or not arg:
            raise ValueError(f"{op} requires a nonempty identifier")
    elif kind == "choice" and arg not in choices:
        raise ValueError(f"{op} must be one of {choices}")


def evaluate_condition(value: dict, context: dict) -> bool | None:
    """Three-valued evaluation against a playthrough context.

    Context keys: numeric (`level`, `happiness`, ...) and collections keyed by
    the leaf op (for example `has_pokemon: {"ralts"}`). `trade`, `overworld_rain`
    and `device_upside_down` read boolean context keys of the same name.
    """
    validate_condition(value)
    op = value["op"]
    if op == "unknown":
        return None
    if op == "always":
        return True
    if op in {"and", "or"}:
        results = [evaluate_condition(arg, context) for arg in value["args"]]
        decisive = op == "or"
        if decisive in results:
            return decisive
        return None if None in results else not decisive
    if op in _NUMERIC_CONTEXT:
        key = _NUMERIC_CONTEXT[op]
        return None if key not in context else context[key] >= value["value"]
    if op not in context:
        return None
    if _LEAVES[op][0] == "none":
        return bool(context[op])
    return value["value"] in context[op]


def unknown(reason: str) -> dict:
    return {"op": "unknown", "reason": reason}


def all_of(*args: dict) -> dict:
    flat = [a for a in args if a != {"op": "always"}]
    if not flat:
        return {"op": "always"}
    return flat[0] if len(flat) == 1 else {"op": "and", "args": flat}

File: web/src/domain/conditions.ts
/**
 * Phrasing for the typed condition AST.
 *
 * One entry per operator in `rotom_dex/domain/conditions.py`. An operator this build does not know
 * still renders, verbatim — omitting an unrecognised prerequisite would make a route look easier
 * than it is, which is the one mistake this product must not make.
 *
 * `unknown` reasons are never paraphrased: they are the project's own abstention text.
 */

import type { Condition, ConditionLeaf } from '@/api/types'

export type Phrase = { text: string; value?: string; kind: 'requirement' | 'unknown' | 'always' }

const TITLE_EXCEPTIONS: Record<string, string> = { tm: 'TM', hm: 'HM', tr: 'TR', pp: 'PP', hp: 'HP' }

export function titleise(slug: string | number | undefined | null): string {
  if (slug === undefined || slug === null) return ''
  return String(slug)
    .split(/[-_]/)
    .map((part) => TITLE_EXCEPTIONS[part] ?? part.charAt(0).toUpperCase() + part.slice(1))
    .join(' ')
}

const LEAF_PHRASES: Record<string, (value: string) => string> = {
  always: () => 'No prerequisite recorded',
  trade: () => 'Must be traded',
  overworld_rain: () => 'While it is raining in the overworld',
  device_upside_down: () => 'With the console held upside down',
  level_at_least: (v) => `At least level ${v}`,
  happiness_at_least: (v) => `Friendship at least ${v}`,
  beauty_at_least: (v) => `Beauty at least ${v}`,
  affection_at_least: (v) => `Affection at least ${v}`,
  has_pokemon: (v) => `You have ${titleise(v)}`,
  has_item: (v) => `You have ${titleise(v)}`,
  use_item: (v) => `Use ${titleise(v)} on it`,
  held_item: (v) => `Holding ${titleise(v)}`,
  knows_move: (v) => `Knows ${titleise(v)}`,
  knows_move_type: (v) => `Knows a ${titleise(v)}-type move`,
  party_has_pokemon: (v) => `${titleise(v)} is in your party`,
  party_has_type: (v) => `A ${titleise(v)}-type is in your party`,
  trade_for_pokemon: (v) => `Trade for ${titleise(v)}`,
  at_location: (v) => `At ${titleise(v)}`,
  in_region: (v) => `In ${titleise(v)}`,
  milestone: (v) => `After ${titleise(v)}`,
  encounter_condition: (v) => `Encounter condition: ${titleise(v)}`,
  encounter_pokemon: (v) => `After encountering ${titleise(v)}`,
  time_of_day: (v) => `During ${v}`,
  gender: (v) => `${titleise(v)} only`,
  stat_relation: (v) => `Attack and Defense: ${v}`,
}

export function phraseLeaf(leaf: ConditionLeaf): Phrase {
  if (leaf.op === 'unknown') {
    return { text: leaf.reason ?? 'Not determined.', kind: 'unknown' }
  }
  if (leaf.op === 'always') return { text: LEAF_PHRASES.always!(''), kind: 'always' }
  const render = LEAF_PHRASES[leaf.op]
  const value = leaf.value === undefined ? '' : String(leaf.value)
  if (!render) {
    // Unknown to this build, but real to the server. Show it plainly rather than hide it.
    return { text: `${titleise(leaf.op)}${value ? `: ${titleise(value)}` : ''}`, value, kind: 'requirement' }
  }
  return { text: render(value), value, kind: leaf.op === 'always' ? 'always' : 'requirement' }
}

export function isCombinator(condition: Condition): condition is { op: 'and' | 'or'; args: Condition[] } {
  return (condition.op === 'and' || condition.op === 'or') && Array.isArray((condition as { args?: unknown }).args)
}

export function isTrivial(condition: Condition | null | undefined): boolean {
  return !condition || condition.op === 'always'
}

/** Flat list of every leaf, for summaries and counts. */
export function flattenLeaves(condition: Condition): ConditionLeaf[] {
  if (isCombinator(condition)) return condition.args.flatMap(flattenLeaves)
  return [condition as ConditionLeaf]
}

export function countUnknown(condition: Condition | null | undefined): number {
  if (!condition) return 0
  return flattenLeaves(condition).filter((leaf) => leaf.op === 'unknown').length
}

export const VERDICT_LABEL: Record<string, string> = {
  reachable: 'Reachable',
  locked: 'Blocked',
  unknown: 'Not determined',
  'not-applicable': 'Not in this game',
}

export const VERDICT_EXPLANATION: Record<string, string> = {
  reachable: 'Every prerequisite you have vouched for is met.',
  locked: 'Something you have recorded as complete is missing a prerequisite.',
  unknown: 'Rotom cannot tell from what has been reviewed and what you have recorded. This is not a claim that it is unobtainable.',
  'not-applicable': 'This rule cannot fire in this game.',
}
